### ISSUE TITLE
habitatmap_terr_2020_v1: the 'habitatmap_terr_types' element violates the primary key

### ISSUE THREAD

**Issue by florisvdh** [inbo/n2khab-preprocessing#66]
While reading the conversations from 2021 in PR https://github.com/inbo/n2khab-preprocessing/pull/50, the statement that the 'xxx_types' elements of `habitatmap_stdized` and `habitatmap_terr`, versions 2020, have a primary key on `polygon_id` + `type` + `certain` piqued my interest. It is also stated in the documentation of `n2khab::read_habitatmap_stdized()` and `n2khab::read_habitatmap_terr()`.

I wanted to check these statements for myself, and found this.

- the requirement is fulfilled for `habitatmap_stdized`
- however there's one violation for `habitatmap_terr`, see below. (Have not yet looked into the proposed 2023 versions of https://github.com/inbo/n2khab-preprocessing/pull/65)

``` r
library(n2khab)
#> Attaching n2khab 0.10.1.9000.
#> Will use sf 1.0-15.
#> Will use terra 1.7-72.
library(dplyr, warn.conflicts = FALSE)

# to verify this is version habitat_stdized_2020_v1
xxh64sum(file.path(locate_n2khab_data(), "20_processed/habitatmap_stdized/habitatmap_stdized.gpkg"))
#> habitatmap_stdized.gpkg 
#>      "3109c26f0a27a0f3"
# to verify this is version habitatmap_terr_2020_v1
xxh64sum(file.path(locate_n2khab_data(), "20_processed/habitatmap_terr/habitatmap_terr.gpkg"))
#> habitatmap_terr.gpkg 
#>   "72d3144016e816ed"

# habitatmap_stdized

hmstd <- read_habitatmap_stdized()
hmstd$habitatmap_types |> count(polygon_id, type, certain) |> filter(n > 1)
#> # A tibble: 0 × 4
#> # ℹ 4 variables: polygon_id <fct>, type <fct>, certain <lgl>, n <int>

# habitatmap_terr

hmt <- read_habitatmap_terr()
hmt$habitatmap_terr_types |> count(polygon_id, type, certain) |> filter(n > 1)
#> # A tibble: 1 × 4
#>   polygon_id   type      certain     n
#>   <fct>        <fct>     <lgl>   <int>
#> 1 427121_v2020 7140_meso TRUE        2
hmt$habitatmap_terr_types |> filter(polygon_id == "427121_v2020")
#> # A tibble: 2 × 6
#>   polygon_id   type      certain code_orig  phab source                         
#>   <fct>        <fct>     <lgl>   <chr>     <int> <fct>                          
#> 1 427121_v2020 7140_meso TRUE    7140         95 habitatmap_stdized + interpret…
#> 2 427121_v2020 7140_meso TRUE    7140_meso     5 habitatmap_stdized
```

<sup>Created on 2024-03-11 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.3.3 (2024-02-29)
#>  os       Linux Mint 21.3
#>  system   x86_64, linux-gnu
#>  ui       X11
#>  language nl_BE:nl
#>  collate  nl_BE.UTF-8
#>  ctype    nl_BE.UTF-8
#>  tz       Europe/Brussels
#>  date     2024-03-11
#>  pandoc   3.1.1 @ /usr/lib/rstudio/resources/app/bin/quarto/bin/tools/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version     date (UTC) lib source
#>  assertthat    0.2.1       2019-03-21 [3] CRAN (R 4.0.1)
#>  class         7.3-22      2023-05-03 [3] RSPM (R 4.2.0)
#>  classInt      0.4-10      2023-09-05 [3] RSPM (R 4.3.0)
#>  cli           3.6.2       2023-12-11 [3] RSPM (R 4.3.0)
#>  curl          5.2.1       2024-03-01 [3] RSPM (R 4.3.0)
#>  DBI           1.2.2       2024-02-16 [3] RSPM (R 4.3.0)
#>  digest        0.6.34      2024-01-11 [3] RSPM (R 4.3.0)
#>  dplyr       * 1.1.4       2023-11-17 [3] RSPM (R 4.3.0)
#>  e1071         1.7-14      2023-12-06 [3] RSPM (R 4.3.0)
#>  evaluate      0.23        2023-11-01 [3] RSPM (R 4.3.0)
#>  fansi         1.0.6       2023-12-08 [3] RSPM (R 4.3.0)
#>  fastmap       1.1.1       2023-02-24 [3] RSPM (R 4.2.0)
#>  forcats       1.0.0       2023-01-29 [3] RSPM (R 4.2.0)
#>  fs            1.6.3       2023-07-20 [3] RSPM (R 4.2.0)
#>  generics      0.1.3       2022-07-05 [3] RSPM (R 4.2.0)
#>  git2r         0.33.0      2023-11-26 [3] RSPM (R 4.3.0)
#>  git2rdata     0.4.0       2022-03-17 [3] RSPM (R 4.2.0)
#>  glue          1.7.0       2024-01-09 [3] RSPM (R 4.3.0)
#>  htmltools     0.5.7       2023-11-03 [3] RSPM (R 4.3.0)
#>  KernSmooth    2.23-22     2023-07-10 [3] RSPM (R 4.2.0)
#>  knitr         1.45        2023-10-30 [3] RSPM (R 4.3.0)
#>  lifecycle     1.0.4       2023-11-07 [3] RSPM (R 4.3.0)
#>  magrittr      2.0.3       2022-03-30 [3] RSPM (R 4.2.0)
#>  n2khab      * 0.10.1.9000 2024-01-24 [1] local
#>  pillar        1.9.0       2023-03-22 [3] RSPM (R 4.2.0)
#>  pkgconfig     2.0.3       2019-09-22 [3] CRAN (R 4.0.1)
#>  plyr          1.8.9       2023-10-02 [3] RSPM (R 4.3.0)
#>  proxy         0.4-27      2022-06-09 [3] RSPM (R 4.2.0)
#>  purrr         1.0.2       2023-08-10 [3] RSPM (R 4.2.0)
#>  R.cache       0.16.0      2022-07-21 [3] RSPM (R 4.2.0)
#>  R.methodsS3   1.8.2       2022-06-13 [3] RSPM (R 4.2.0)
#>  R.oo          1.26.0      2024-01-24 [3] RSPM (R 4.3.0)
#>  R.utils       2.12.3      2023-11-18 [3] RSPM (R 4.3.0)
#>  R6            2.5.1       2021-08-19 [3] RSPM (R 4.2.0)
#>  Rcpp          1.0.12      2024-01-09 [3] RSPM (R 4.3.0)
#>  remotes       2.4.2.1     2023-07-18 [3] RSPM (R 4.2.0)
#>  reprex        2.1.0       2024-01-11 [3] RSPM (R 4.3.0)
#>  rlang         1.1.3       2024-01-10 [3] RSPM (R 4.3.0)
#>  rmarkdown     2.26        2024-03-05 [3] RSPM (R 4.3.0)
#>  rprojroot     2.0.4       2023-11-05 [3] RSPM (R 4.3.0)
#>  rstudioapi    0.15.0      2023-07-07 [3] RSPM (R 4.2.0)
#>  sessioninfo   1.2.2       2021-12-06 [3] RSPM (R 4.2.0)
#>  sf            1.0-15      2023-12-18 [1] RSPM (R 4.3.2)
#>  stringi       1.8.3       2023-12-11 [3] RSPM (R 4.3.0)
#>  stringr       1.5.1       2023-11-14 [3] RSPM (R 4.3.0)
#>  styler        1.10.2      2023-08-29 [3] RSPM (R 4.2.0)
#>  tibble        3.2.1       2023-03-20 [3] RSPM (R 4.3.0)
#>  tidyr         1.3.1       2024-01-24 [3] RSPM (R 4.3.0)
#>  tidyselect    1.2.0       2022-10-10 [3] RSPM (R 4.2.0)
#>  units         0.8-5       2023-11-28 [3] RSPM (R 4.3.0)
#>  utf8          1.2.4       2023-10-22 [3] RSPM (R 4.3.0)
#>  vctrs         0.6.5       2023-12-01 [3] RSPM (R 4.3.0)
#>  withr         3.0.0       2024-01-16 [3] RSPM (R 4.3.2)
#>  xfun          0.42        2024-02-08 [3] RSPM (R 4.3.0)
#>  yaml          2.3.8       2023-12-11 [3] RSPM (R 4.3.0)
#> 
#>  [1] /home/floris/lib/R/library
#>  [2] /usr/local/lib/R/site-library
#>  [3] /usr/lib/R/site-library
#>  [4] /usr/lib/R/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>

It is quite clear what is the difference between the two rows; the clash in the primary keys is due to the further interpretation of 7140 as 7140_meso, while the latter is also a code that already existed for the polygon.

By definition, 'certain' in the `habitatmap_terr` data source is the value of 'certain' in `habitatmap_stdized`, so it does not apply to the interpretation.

I think of multiple possible ways to go forward with this (including: updating primary key statement, not the data source), but since I didn't yet recap everything about the creation of these data sources I should stop here for now.

Opinions on this @ToonHub @cecileherr @w-jan?

**Comment by florisvdh** [comment]
Adding that `n2khab::read_habitatmap_stdized()` also documents the following:

> After those first two steps, a given polygon could contain the same type with the same value for certain repeated several times, e.g. when 31xx_rbbmr is present with phab = yy% and 31xx is present with phab = zz%. In that case the rows with the same polygon_id, type and certain were **gathered into one row** and the respective phab values were added up.

**Comment by florisvdh** [comment]
The creation of `habitatmap_terr` involves filtering `habitatmap_stdized` and the modification of specific type codes. However it does not evaluate the need for another aggregation in order to meet the primary key condition, in the way it has been implemented in `habitatmap_stdized` as below (using a semicolon to collapse original codes):

https://github.com/inbo/n2khab-preprocessing/blob/0b94cd25a76f77bd802dbc65cd4ebe3fb7fb4a5e/src/generate_habitatmap_stdized/10_generate_habmap_stdized.Rmd#L195-L203

I would suggest to do this aggregation in `habitatmap_terr` though, in order to respect the stated primary key where interpretation has led to duplicate key values.

I think an update of the `habitatmap_terr_2020_v1` would indeed be useful (creating a v2 version), since that version was already intended to have the primary key.

Anyway it's not worth the hassle to update the 2018_v2 versions, where the primary key was not there for sure (57 duplicated records in `habitatmap_stdized`, 13 in `habitatmap_terr`).

BTW with modern {dplyr}, above referred code for aggregation can be simplified as:

```r
habmap_long_aggr <- habmap_long %>%
  summarise(
    phab = sum(phab),
    code_orig = str_c(code_orig, collapse = "; "),
    certain = all(certain),
    .by = c(polygon_id, type, certain)
  )
```

The proposed upcoming version of `habitatmap_terr` (derived from `habitatmap_2023`; #65) does not violate the primary key though, so until now (and neglecting 2018) the absence of aggregation only affects 2020_v1.

**Comment by florisvdh** [comment]
This was implemented by @cecileherr in PR #69 :rocket:!
